Fido2 follow up

Ollrogge opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md.
-->

### Contribution description
This PR adjusts the FIDO2 CTAP flash handling to reserve flash memory at compile time using a `static const` buffer. This ensures that the FIDO2 CTAP implementation does not interfere with other modules wanting to use flash memory and doesn't corrupt firmware or e.g. bootloaders at the end of flash. Furthermore configuring `CONFIG_FIDO2_CTAP_NUM_FLASHPAGES` such that more flash than available is required will fail at build time rather than runtime.

<!--
Put here the description of your contribution:
- describe which part(s) of RIOT is (are) involved
- if it's a bug fix, describe the bug that it solves and how it is solved
- you can also give more information to reviewers about how to test your changes
-->


### Testing procedure

<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->
*  `tests/sys_fido2_ctap`

I successfully ran the tests on the `nrf52840dongle` and `nrf52840dk`.

PeterKietzmann commented on sys/fido2/ctap/ctap_mem.c:
```diff
@@ -154,3 +166,18 @@ int fido2_ctap_mem_get_offset_of_rk_into_flashpage(uint16_t rk_idx)
+
+unsigned fido2_ctap_mem_get_state_flashpage_number(void)
+{
+#if defined(CONFIG_FIDO2_CTAP_NUM_FLASHPAGES) && \
```
Should this be `CONFIG_FIDO2_CTAP_FLASH_NUM_FLASHPAGES` (generated from [here](https://github.com/RIOT-OS/RIOT/pull/17096/files#diff-2c0cb8c9ff04ebbb2b2d14a8c59f79e75e94fad8e56e96dcc1555a32147cd614R87))? Btw, prefer the name `CONFIG_FIDO2_CTAP_NUM_FLASHPAGES`.

PeterKietzmann commented on sys/include/fido2/ctap/ctap_mem.h:
```diff
@@ -41,28 +41,9 @@ extern "C" {
+ * @brief   Default amount of flashpages to use
  */
-#define CTAP_FLASH_STATE_PAGE CTAP_FLASH_RK_START_PAGE - 1
+#define CTAP_FLASH_DEFAULT_AMOUNT_FLASHPAGES 4
```
The default value of `FIDO2_CTAP_FLASH_NUM_FLASHPAGES` is already 4. Do we need the additional `#define`

PeterKietzmann commented on sys/fido2/ctap/Kconfig:
```diff
@@ -84,12 +84,13 @@ config FIDO2_CTAP_UP_BUTTON_FLANK_RISING
-        Configuring this incorrectly can lead to firmware corruption so make sure
-        the flash page is located after the firmware.
+        Configure how many flashpages are used to store FIDO2 CTAP data.
+        Configuring this incorrectly can lead to firmware corruption.
```
In which cases? Can you give more detail here?

PeterKietzmann commented on sys/include/fido2/ctap/ctap_mem.h:
```diff
@@ -43,7 +43,12 @@ extern "C" {
+#define CTAP_FLASH_NUM_FLASHPAGES CONFIG_FIDO2_CTAP_NUM_FLASHPAGES
+#else
+#define CTAP_FLASH_NUM_FLASHPAGES 4
+#endif
```
How about:

```
#ifndef CONFIG_FIDO2_CTAP_NUM_FLASHPAGES
#define CONFIG_FIDO2_CTAP_NUM_FLASHPAGES 4
#endif

#if CONFIG_FIDO2_CTAP_NUM_FLASHPAGES <= 2 || CONFIG_FIDO2_CTAP_NUM_FLASHPAGES >= 256
#error "why failure?"
#endif
```
(haven't tested it)

This way you would get rid of the extra symbol.

Ollrogge replied:
This would also work. I would change `CONFIG_FIDO2_CTAP_NUM_FLASHPAGES >= 256` to `CONFIG_FIDO2_CTAP_NUM_FLASHPAGES >= FLASHPAGE_NUMOF` though.

Overall it is hard to give an upper bound for the max number of flashpages CTAP can obtain ...

PeterKietzmann replied:
Sounds good, please go ahead

PeterKietzmann commented:
Tested with a nrf52840dk:
- Compiled with default number of flashpages -> got 4.
- Compiled with 8 flashpages -> got 8.
- Compiled with `CFLAGS=-DCONFIG_FIDO2_CTAP_NUM_FLASHPAGES=1 and 257` -> fails(expected).
- Unit tests succeed `126 passed, 7 skipped, 7 deselected in 149.62s (0:02:29)`.
- Registration and login with 4 accounts still works on *webauthn.io*.

PeterKietzmann commented:
The changes look good in general, but I think the documentation can be improved. From a user perspective: I can now change the number of flashpages (jippie!) but what does that mean, how should I configure that value? For example, I know that I want to serve 12 accounts with my authenticator, how many flashpages do I need?

benpicco commented on sys/fido2/ctap/ctap_mem.c:
```diff
@@ -154,3 +161,13 @@ int fido2_ctap_mem_get_offset_of_rk_into_flashpage(uint16_t rk_idx)
+
+unsigned fido2_ctap_mem_get_state_flashpage_number(void)
+{
+    return flashpage_last_free() - CONFIG_FIDO2_CTAP_NUM_FLASHPAGES + 1;
```
Are we sure that this will be below `flashpage_first_free()`?

Ollrogge replied:
No. Should I add a check into [ctap_mem_init](https://github.com/RIOT-OS/RIOT/blob/5e925e1906a7d10b2ab6f8d9b972755b8d290d19/sys/fido2/ctap/ctap_mem.c#L48) checking this condition and returning an error if it doesn't hold ?

PeterKietzmann replied:
Side question: Why do follow the logic of writing to the end of the flash at all? Why not use first free?

Ollrogge replied:
Because `flashpage_first_free()` will return a different value across firmware updates if riotboot is used in two-slot mode. This could result in the loss of FIDO2 data.

benpicco replied:
But `flashpage_last_free()` also returns a different value across firmware updates (always the end of the current slot)

Ollrogge replied:
Oh really ? Then I misunderstood how the riotboot slots work. From the [documentation](https://api.riot-os.org/group__bootloader__riotboot.html) I would have thought that slot 0 and slot 1 are at the beginning of the flash after the bootloader and the rest of the flash is unused. In this case `flashpage_last_free()` would always be the same.

benpicco replied:
With riotboot the flash is split in two halves, the first half is slot 0 and the second half is slot1.

You can't have the slots have dynamic sizes as otherwise a firmware update would not be allowed to be different in size than the previous version.

btw: You added [a warning](https://github.com/RIOT-OS/RIOT/blob/master/drivers/include/periph/flashpage.h#L225) to `flashpage_last_free()` yourself :wink:

Ollrogge replied:
Thanks for the clarification !

Ollrogge replied:
So should I change it to use `flashpage_first_free()`? And what about the case where `flashpage_first_free()` is not above `flashpage_first_free()` ?

PeterKietzmann replied:
If I understand it correctly, neither `flashpage_first_free` nor `flashpage_last_free` works in the case of firmware updates. Is that correct?
- Can we reserve a flash region prior to assigning slot 0/1? There are other use cases with such requirements (#17092).
- Side note: FIDO2 keys that are currently stored in flash will likely go to a secure element in future (WIP). So the remaining data to be stored in flash is only configurations, I suppose.
- For the FIDO2 case: Do we actually need to consider firmware updates? How would a deployment look like? Otherwise, we can ignore this problem and simply document it properly.

Ollrogge replied:
Fixed the code to use `flashpage_first_free()`.
I agree with @PeterKietzmann that firmware updates should not be considered as of now.

benpicco commented:
A bit unrelated, but what made you chose TinyCBOR instead of NanoCBOR for CBOR parsing?
NanoCBOR is the default in RIOT (and used for SUIT update), so if you really enable firmware updates you'll end up with two CBOR libraries linked in.

Ollrogge commented:
> A bit unrelated, but what made you chose TinyCBOR instead of NanoCBOR for CBOR parsing? NanoCBOR is the default in RIOT (and used for SUIT update), so if you really enable firmware updates you'll end up with two CBOR libraries linked in.

Honestly I can't remember. This decision was made really early when I barely had any experience with RIOT. I think someone recommended TinyCBOR and I just went with it

chrysn commented on sys/fido2/ctap/Kconfig:
```diff
@@ -84,12 +84,22 @@ config FIDO2_CTAP_UP_BUTTON_FLANK_RISING
+        To save authenticator state data one additional flashpage is needed.
+        So in total one has to configure 3 to save 40 credentials.
+
+        Configuring the amount too high can lead to loss of FIDO2 data as a
```
Why not check against `flashpage_last_free()`?

Could probably not be caught at Kconfig time, and maybe not even at build time, but a runtime error ("Sorry can't use my flash" or "You only get this much flash rather than what you configured; sorry"), but all that would still be better than silent failure.

chrysn replied:
Just saw that this is already being discussed somewhere else in the code; please mark as resolved once that's done.

chrysn commented:
The way `flashpage_{first,last}_free` is used here begs clarification from their documentation.

If I came to these functions without the context of the present PR, I'd assume that these "free" pages are free for the application to use.

Here, something internal to RIOT uses them -- and my expectation of a modules system is that I can enable any additional modules and my application would still run unimpeded.

A possible way out here is for the flashpage_*_free mechanism would be to keep track of that some system component has "claimed" a chunk of flash (and because that happens in ordered start code, they get the same pages across reboots) -- but that would be doing a check at runtime that could also be done at built time.

(In own applications, I've always allocated flash like

```c
static const uint8_t backing_memory[FLASHPAGE_SIZE] __attribute__((aligned(FLASHPAGE_SIZE))) = {0};

...
    int flashpage = flashpage_page((void*)backing_memory);
```

which cooperates nicely with the flashpage_*_free mechanism).

Ollrogge commented:
@chrysn thanks for the suggestions ! I think your suggested aproach is the best solution. I updated the code accordingly.

PeterKietzmann commented:
> @chrysn thanks for the suggestions ! I think your suggested aproach is the best solution. I updated the code accordingly.

Does this still interfere with the firmware update case?

@Ollrogge can you update the documentation?

Ollrogge commented:
> > @chrysn thanks for the suggestions ! I think your suggested aproach is the best solution. I updated the code accordingly.
> 
> Does this still interfere with the firmware update case?

To my understanding yes. This does still interfere with the firmware update case.
> 
> @Ollrogge can you update the documentation?

Which documentation do you think should be updated ?

PeterKietzmann commented:
In *sys/fido2/doc.txt* a note on 1. how flash is allocated and 2. firmware updates are not possible with our fido2 implementation for thisandthat reason. @chrysn other things to document?

chrysn commented:
> @chrysn other things to document?

Not in here, now that it's not using flashpage_*_free any more.

If the changes here removed the necessity to have flashpage_first_free
and flashpage_last_free altogether (which AIU was prepared in a first
stage of this PR), it may be worth considering deprecating (or removing,
not having been in a release) them right away again, or documenting how
they are supposed to be used, but there's nothing in this PR any more
that makes that belong here.

Ollrogge commented:
> @chrysn other things to document?
> Not in here, now that it's not using flashpage_*_free any more. If the changes here removed the necessity to have flashpage_first_free and flashpage_last_free altogether (which AIU was prepared in a first stage of this PR), it may be worth considering deprecating (or removing, not having been in a release) them right away again, or documenting how they are supposed to be used, but there's nothing in this PR any more that makes that belong here.

They did but don't you think the `flashpage_*_free` functions can be helpful for applications wanting to write to flash memory without corrupting e.g. firmware / bootloaders ?

chrysn commented:
> They did but don't you think the `flashpage_*_free` functions can be
> helpful for applications wanting to write to flash memory without
> corrupting e.g. firmware / bootloaders ?

Thing is, they can be used by *only one* component in the whole system
(so the application better not be modular, or later split into library
components).

IMO we should rather provide assistance (be it in examples or even
convenience functions / macros) for applications to create their own
linker-managed area of flash, which can be managed in the isolation of a
compilation unit, shows up in the debug symbols, and screams at build
time if the requested space is not available, rather than hope that the
application manages the space manually based on them.

(There can be a case for them if the application should use "the rest"
of the flash memory, but if this or another use case does justify having
the functions around, I think it should be advertised for the cases, and
point users to the existing possibilities of using the linker).

PeterKietzmann commented on sys/fido2/doc.txt:
```diff
@@ -17,6 +17,9 @@
  * @warning The FIDO2 implementation currently stores private keys in plain text inside flash memory.
  *
+ * @warning This implementation does not support firmware updates if riotboot is used in
+ *          two slot mode since the firmware slots alternate.
```
This does not explain why firmware updates are not supported.

chrysn commented on sys/fido2/doc.txt:
```diff
@@ -17,6 +17,12 @@
  *
  * @warning The FIDO2 implementation currently stores private keys in plain text inside flash memory.
  *
+ * @warning This implementation does not support firmware updates if riotboot is used in
```
From my understanding this could be:

> @warning This implementation only persists FIDO CTAP data across reboots and when unpowered; any firmware update loses the data. This applies both to firmware updates through bootloaders and to firmware updates through external programmers.

Ollrogge commented:
Hey @chrysn, your proposed changes with the `static const` buffer work fine on the `nrf52840dk` but not on the `nrf52840dongle`.
After flashing the dongle everything works fine until I power cycle the dongle. Then it automatically enters bootloader mode again.
Beforehand the dongle would simply start the FIDO2 CTAP application again and everything worked fine.

Do you have any idea what could cause this ? The flashpage returned by `flashpage_page((void *)_backing_memory)` seems to be correct, yet it doesn't work. I am really scratching my head with this one.

Ollrogge commented:
> Hey @chrysn, your proposed changes with the `static const` buffer work fine on the `nrf52840dk` but not on the `nrf52840dongle`. After flashing the dongle everything works fine until I power cycle the dongle. Then it automatically enters bootloader mode again. Beforehand the dongle would simply start the FIDO2 CTAP application again and everything worked fine.
> 
> Do you have any idea what could cause this ? The flashpage returned by `flashpage_page((void *)_backing_memory)` seems to be correct, yet it doesn't work. I am really scratching my head with this one.

A follow up on this. As it turns out the reason for the code to be working on the `nrf52840dk` but not on the `nrf52840dongle` is the [Open USB bootloader](https://devzone.nordicsemi.com/guides/short-range-guides/b/getting-started/posts/nrf52840-dongle-programming-tutorial) that ships with the dongle. Going through the bootloader code located in the [nRF5 SDK](https://www.nordicsemi.com/Products/Development-software/nrf5-sdk/download) I found that in the function `nrf_bootloader_app_start_final` following line is executed: `nrf_bootloader_flash_protect(0, nrf_dfu_bank0_start_addr() + s_dfu_settings.bank_0.image_size, false);`. This write-protects the flash memory of the application. This includes the flash memory reserved by using a `static const buffer`. Therefore when trying to write to this memory an exception is caused and the dongle goes into bootloader mode. 

To my understanding, a linker-managed area of flash as you suggested @chrysn would also not solve this problem. I don't really know how to continue from here.

What are your suggestions / thoughts @chrysn @benpicco @PeterKietzmann ?

Ollrogge commented:
With #17436 being merged I updated this PR to use the newly introduced `FLASH_WRITABLE_INIT` macro. 
Successfully ran the tests on the `nrf52840dk`.

@chrysn This should be fine now as well?

chrysn commented on sys/include/fido2/ctap/ctap_mem.h:
```diff
@@ -41,28 +41,16 @@ extern "C" {
- */
-#define CTAP_FLASH_STATE_PAGE CTAP_FLASH_RK_START_PAGE - 1
+#if CONFIG_FIDO2_CTAP_NUM_FLASHPAGES <= 2 || \
+    CONFIG_FIDO2_CTAP_NUM_FLASHPAGES >= FLASHPAGE_NUMOF
```
The second half of this check is unnecessary; the linker script takes care of this in better granularity. (Especially, it detects the case when the requested memory would fit the flash, but not simultaneously with all the rest of the program).

chrysn commented on sys/fido2/doc.txt:
```diff
@@ -102,6 +106,9 @@
  * Abstraction for flash operations. Uses the RIOT [Flashpage MTD driver](http://api.riot-os.org/group__drivers__mtd__flashpage.html).
  *
+ * Flash memory is reserved at build time using the `FLASH_WRITABLE_INIT` macro. The amount of flash
+ * reserved can be configured via KConfig.
```
```suggestion
 * reserved can be configured as `FIDO2_CTAP_NUM_FLASHPAGES` in KConfig.
```

chrysn commented:
Thanks, my concerns have been addressed. Can't do the full review right now, maybe @PeterKietzmann you could have another look?

PeterKietzmann commented on sys/fido2/ctap/ctap_mem.c:
```diff
@@ -154,3 +166,13 @@ int fido2_ctap_mem_get_offset_of_rk_into_flashpage(uint16_t rk_idx)
+    return flashpage_page((void *)_backing_memory);
+}
+
+unsigned fido2_ctap_mem_get_rk_start_page(void)
```
You want this function explicitly to have an abstraction `+1` (rk flashpage offset)? Does it need to be exposed or would static inline do?

PeterKietzmann commented on sys/fido2/doc.txt:
```diff
@@ -17,6 +17,10 @@
  *
  * @warning The FIDO2 implementation currently stores private keys in plain text inside flash memory.
  *
+ * @warning This implementation only persists FIDO CTAP data across reboots and when unpowered;
```
"only" can be removed

PeterKietzmann commented on sys/fido2/doc.txt:
```diff
@@ -17,6 +17,10 @@
  *
+ * @warning This implementation only persists FIDO CTAP data across reboots and when unpowered;
+ *          any firmware update loses the data. This applies both to firmware updates through
+ *          bootloaders and to firmware updates through external programmers.
```
Maybe add a note why (flash overwritten?)

PeterKietzmann commented on sys/fido2/ctap/ctap.c:
```diff
@@ -1792,12 +1793,15 @@ static int _ctap_decrypt_rk(ctap_resident_key_t *rk, ctap_cred_id_t *id)
 static int _write_state_to_flash(const ctap_state_t *state)
 {
-    return fido2_ctap_mem_write(state, CTAP_FLASH_STATE_PAGE, 0, CTAP_FLASH_STATE_SZ);
+    return fido2_ctap_mem_write(state,
```
If I get it correctly, the "first" (available) page is always used for the state. Would be nice to leave a comment about that here and in `_read_state_from_flash`

PeterKietzmann commented on sys/fido2/doc.txt:
```diff
@@ -102,6 +106,9 @@
  * Abstraction for flash operations. Uses the RIOT [Flashpage MTD driver](http://api.riot-os.org/group__drivers__mtd__flashpage.html).
  *
+ * Flash memory is reserved at build time using the `FLASH_WRITABLE_INIT` macro. The amount of flash
+ * reserved can be configured as `FIDO2_CTAP_NUM_FLASHPAGES` in KConfig.
```
Similarly to the Kconfig help, it would be nice to explain here that 1. page is always state, and subsequent pages are used to store (multiple) rks.

PeterKietzmann commented on sys/fido2/ctap/Kconfig:
```diff
@@ -108,12 +110,19 @@ config FIDO2_CTAP_UP_BUTTON_FLANK_RISING
-        the flash page is located after the firmware.
+        Configure how many flashpages are used to store FIDO2 CTAP data.
+
+        To save a credential in flash memory, roughly 156 bytes are needed. This
```
Try to use the same language here compared to the code. rk versus state versus credential.

PeterKietzmann commented on sys/fido2/ctap/ctap.c:
```diff
@@ -1792,12 +1793,15 @@ static int _ctap_decrypt_rk(ctap_resident_key_t *rk, ctap_cred_id_t *id)
+                               sizeof(*state));
 }
 
 static int _write_rk_to_flash(const ctap_resident_key_t *rk, int page, int offset)
```
This indirection is not needed, is it? It seems to me that you want to have `_save_rk` here (the only user of this function). But I would keep the name `_write_rk_to_flash` to be comparable to `_write_state_to_flash`.

PeterKietzmann commented on sys/include/fido2/ctap/ctap_mem.h:
```diff
@@ -155,6 +142,20 @@ int fido2_ctap_mem_get_flashpage_number_of_rk(uint16_t rk_idx);
+ * @return page number
+ */
+unsigned fido2_ctap_mem_get_rk_start_page(void);
+
```
How about `get_state_flashpage` and `get_rk_flashpage` to be a bit more consistent?

PeterKietzmann reviewed (commented):
@Ollrogge the diff looks good to me, in general. Hence, my comments mostly request for clarification in the documentation. I find it a bit hard to follow the memory related code, though. Can you try to harmonize naming for functions and documentation a bit? Just some indications:
- descriptions for rk use: start, stop, offset (into), page, amount, number index
-> try to use the same couple of terms everywhere.
- page vs flashpage
- credentials vs rk (or is it credentials: rk + state, then it does not get clear)

I will run some tests on the cleaned ve3rsion of this PR.

PeterKietzmann commented on sys/fido2/ctap/ctap.c:
```diff
@@ -282,7 +272,12 @@ int fido2_ctap_init(void)
 
-    ret = _read_state_from_flash(&_state);
+    /**
+     * CTAP state information is stored at flashpage 0 of the memory area
```
Is it really flashpage 0 or offset zero?

Ollrogge replied:
Both. It is stored at offset 0 of flashpage 0

PeterKietzmann commented on sys/fido2/ctap/ctap_mem.c:
```diff
@@ -154,3 +166,8 @@ int fido2_ctap_mem_get_offset_of_rk_into_flashpage(uint16_t rk_idx)
+unsigned fido2_ctap_mem_flash_page(void)
+{
+    return flashpage_page((void *)_backing_memory);
+}
```
Missing newline

PeterKietzmann commented on sys/fido2/doc.txt:
```diff
@@ -102,6 +106,8 @@
  *
  * Abstraction for flash operations. Uses the RIOT [Flashpage MTD driver](http://api.riot-os.org/group__drivers__mtd__flashpage.html).
  *
+ * Flash memory is reserved at build time using the `FLASH_WRITABLE_INIT` macro. The amount of flashpages reserved can be configured as `FIDO2_CTAP_NUM_FLASHPAGES` in KConfig. The implementation needs atleast 1 flashpage to store state information and 1 flashpage to store credentials (rks). Therefore, the minimum amount of flashpages needed is 2. State information is stored on the first flashpage ...
```
1. This line is too long
2. Can you introduce the abbreviation 'rk' here since it is the first occurance in the documentation?

Ollrogge replied:
Is there a line length limit for doc.txt files ? Running `make doc` does not show any errors and the produced html also looks fine

PeterKietzmann commented on sys/include/fido2/ctap/ctap_mem.h:
```diff
@@ -41,28 +41,15 @@ extern "C" {
- * @brief Page for storing authenticator state information
- */
-#define CTAP_FLASH_STATE_PAGE CTAP_FLASH_RK_START_PAGE - 1
+#if CONFIG_FIDO2_CTAP_NUM_FLASHPAGES <= 2
```
Why smaller **equal**? One for state + at least one for rks? It also contradicts the Kconfig range in this way.

Ollrogge replied:
Correct that is wrong. Ty!

PeterKietzmann commented on sys/fido2/ctap/ctap.c:
```diff
@@ -192,16 +192,6 @@ static int _save_pin(uint8_t *pin, size_t len);
  */
 static int _write_state_to_flash(const ctap_state_t *state);
```
Though it is not directly related to the PR, I still think it would improve the code if you rename `_save_rk` to be in line with this function naming scheme (as [commented before](https://github.com/RIOT-OS/RIOT/pull/17096#discussion_r833110378))

PeterKietzmann reviewed (commented):
I've tested *tests/sys_fido2_ctap* on a nrf52040dk.

- Registration and login on *webauthn.io* works (tested with three accounts), even after a device power off between logins.
- Reconfiguration of flashpages works (despite the 'smaller equal thingy' commented inline).
- `make fido2-test` succeeds

Question: 
I have configured 2 pages (after fixing the 'smaller equal thingy'). According to the [documentation](https://github.com/RIOT-OS/RIOT/pull/17096/files#diff-2c0cb8c9ff04ebbb2b2d14a8c59f79e75e94fad8e56e96dcc1555a32147cd614R123), this should be enough for ~ 4096/156=26 credentials. So I registered multiple accounts (35) on *webauthn.io* and couldn't see any failure. (Why) is that expected? What is the strategy when running out of space?

Ollrogge commented:
> I've tested _tests/sys_fido2_ctap_ on a nrf52040dk.
> 
>     * Registration and login on _webauthn.io_ works (tested with three accounts), even after a device power off between logins.
> 
>     * Reconfiguration of flashpages works (despite the 'smaller equal thingy' commented inline).
> 
>     * `make fido2-test` succeeds
> 
> 
> Question: I have configured 2 pages (after fixing the 'smaller equal thingy'). According to the [documentation](https://github.com/RIOT-OS/RIOT/pull/17096/files#diff-2c0cb8c9ff04ebbb2b2d14a8c59f79e75e94fad8e56e96dcc1555a32147cd614R123), this should be enough for ~ 4096/156=26 credentials. So I registered multiple accounts (35) on _webauthn.io_ and couldn't see any failure. (Why) is that expected? What is the strategy when running out of space?

Thank you very much for testing ! I had forgotten to adjust the check for the maximum amount of resident keys. It should be working now. I ran the same test as you and received an error when trying to create the 27th credential.

PeterKietzmann commented on sys/include/fido2/ctap/ctap_mem.h:
```diff
@@ -84,6 +71,12 @@ extern "C" {
+/**
+ * @brief Max amount of resident keys that can be stored on device
+ */
+#define CTAP_FLASH_MAX_RK_AMT ((CONFIG_FIDO2_CTAP_NUM_FLASHPAGES - 1) * \
```
I would prefer `CTAP_FLASH_MAX_NUM_RKS`

PeterKietzmann reviewed (commented):
> I ran the same test as you and received an error when trying to create the 27th credential.

How did you observe the error? Is there no LED indication for a "key store full" event? Otherwise, a `print` or `DEBUG` would be good. 

Surprisingly, I do not even seem to enter `_write_rk_to_flash` on registration. Can you check that please?

Ollrogge commented:
> 



> > I ran the same test as you and received an error when trying to create the 27th credential.
> 
> How did you observe the error? Is there no LED indication for a "key store full" event? Otherwise, a `print` or `DEBUG` would be good.

I used the [pythin-fido2](https://github.com/Yubico/python-fido2) client and got a `CTAP2_ERR_KEY_STORE_FULL return code.
> 
> Surprisingly, I do not even seem to enter `_write_rk_to_flash` on registration. Can you check that please?

If you are using WebAuthn.io make sure to set advanced settings -> Register with Resident Key -> Required. Otherwise the authenticator will store the credential on the FIDO2 server in encrypted form.

PeterKietzmann commented:
On a positive note, I was able to reproduce the overflow after registering 26 accounts.
![Bildschirmfoto_2022-03-28_10-58-52](https://user-images.githubusercontent.com/7765855/160367467-c3c578e5-a45e-4430-ba15-21df27a8b334.png)

New questions occurred, though. 
- Apparently, I need to use a password now . Why? I think that I chose '1234' initially. It does not change after a power-off cycle *or* re-flash.
- Re-flashing the device **after** registering an account (with 'Register with Resident Key') does not prevent from a successful login with the 'old' account.

Ollrogge commented:
>     * Apparently, I need to use a pin now . Why? I think that I chose '1234' initially. It does not change after a power-off cycle _or_ re-flash.

That you have to use a PIN is caused by WebAuthn.io. Once you chose to save resident keys, they will require you to set a PIN.

> 
>     * Re-flashing the device **after** registering an account (with 'Register with Resident Key') does not prevent from a successful login with the 'old' account.

That is expected. Re-flashing the device will not cause different flashpages to be used. Therefore the content is still there. Using the "reset" command will delete the state information and therefore the authenticator will "forget" about PIN and credentials but the information will still be saved on the flashpages and could be read. Thinking about that, this should be changed. I will implement the erasure of all flashpages when the authenticator receives a RESET command.

That being said I don't think this has anything to do with this PR. How about I open another PR where we talk about general follow-ups for the FIDO2 implementation like better documentation and proper erasure of sensitive data ?

PeterKietzmann commented:
> Using the "reset" command will delete the state information and therefore the authenticator will "forget" about PIN

this contradicts 

>  It does not change after a power-off cycle

Can you check that?

PeterKietzmann commented:
> Using the "reset" command will delete the state information and therefore the authenticator will "forget" about PIN

this contradicts 

>  It does not change after a power-off cycle

Can you check that?

Ollrogge commented:
> > Using the "reset" command will delete the state information and therefore the authenticator will "forget" about PIN
> 
> this contradicts

Why does this contradict ? Bottom line is: If the authenticator receives the reset command it will delete all state information about PIN / credentials. It can also be seen in the [code](https://github.com/RIOT-OS/RIOT/blob/c078fdeb70dd74f29240ff70ebf890e8967fb7aa/sys/fido2/ctap/ctap.c#L433).

> 
> > It does not change after a power-off cycle
> 
> Can you check that?

Why would the PIN need to change after a power cycle ?

PeterKietzmann commented:
Sorry, that was a bit rushed yesterday. I was still referring to the re-flash. So when is the "reset command" executed? To make my observation clearer:
- I initiated a PIN, yesterday.
- I re-flashed my board, now.
- *webauthn.io* registration asks me for a PIN;  that is the one I initiated yesterday (before re-flash).

To me that is unexpected.

Ollrogge commented:
> Sorry, that was a bit rushed yesterday. I was still referring to the re-flash. So when is the "reset command" executed?

Is is executed when you send it :) It is the same kind of [command](https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-client-to-authenticator-protocol-v2.0-id-20180227.html#authenticatorReset) like e.g. `MakeCredential`. 

I think for authenticators you can buy online you need a special client to send the command. In the current FIDO2 implementation in RIOT no protection mechanisms are in place so anyone could send this command.

>  To make my observation clearer:
> 
>     * I initiated a PIN, yesterday.
> 
>     * I re-flashed my board, now.
> 
>     * _webauthn.io_ registration asks me for a PIN;  that is the one I initiated yesterday (before re-flash).
> 
> 
> To me that is unexpected.

Yes I understand why this is unexpected but based on the current implementation it makes sense:
When the authenticator reboots after a re-flash it reads the state information from flash and [checks](https://github.com/RIOT-OS/RIOT/blob/6849e338549473e410dbf3817817e68110117dc8/sys/fido2/ctap/ctap.c#L292) it. Since a re-flash does not erase any flashpages after the firmware all state information will still be there. Therefore the PIN will still work.

Ollrogge commented:
> > Is is executed when you send it :) It is the same kind of [command](https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-client-to-authenticator-protocol-v2.0-id-20180227.html#authenticatorReset) like e.g. MakeCredential.
> 
> "Send it" from where? Is it a feature on _webauthn.io_, an other existing "client" implementation, our unit tests? Or do I have to write my own application to reset? In the latter case I would **at least** expect a little RIOT helper tool for that.
> 
> > Therefore the PIN will still work.
> 
> I understand. My major concern is the absence, or missing documentation on how to reset it (reflashing "any other" application is not a solution).

I added a python script to the test application directory that allows for resetting the authenticator. Also documentated it in the README.
I also added full erasure of all flashpages containing CTAP data when the authenticator receives the reset command.

PeterKietzmann commented:
- `make fido2-up` -> success
- register/login few accounts with residential keys -> success
- power off and login again -> success
- re-flash does not destroy the credentials (as agreed) -> success
- reset (by python script) removes the pin&credentials -> success
- old pin doesn't work anymore -> success
- lowering the number of flashpages to two throws an error after creating > 26 accounts -> success

Can you please verify one aspect (maybe I just did a mistake during monkey.creating several accounts): it seems that after hitting the maximum number of accounts (at 27 accounts with 2 flashpages) old accounts cannot login anymore.
![Bildschirmfoto_2022-04-07_12-18-25](https://user-images.githubusercontent.com/7765855/162182608-07b2c569-7043-4997-88bb-2bcdf96d40ce.png)

PeterKietzmann commented:
Otherwise, the PR looks good and the reset abilities really improve our implementation ;)

Ollrogge commented:
> > Can you please verify one aspect
> 
> A bit more specifically:
> 
>     * Registering 25 accounts (resident key required) works well and all seem to work for login.
> 
>     * Registering account number 26 indicates success **but** after registering it, no login works anymore (like screenshot above).
> 
>     * Registering account number 27 throws the error about out of memory.

Found the issue, thanks for testing so rigorously. The problem was the wrong location of [this](https://github.com/RIOT-OS/RIOT/blob/ad345eea4b0c6c2d408ad22c2d4ac4b7e0c42685/sys/fido2/ctap/ctap.c#L1537) check. `_save_rk` is also called when a rk is updated (e.g. during `get_assertion`). Therefore the check should only be executed if there is actually a new rk to be stored, not if one is just updated.

PeterKietzmann commented:
- `make fido2-test` -> success
- register/login few accounts with residential keys -> success
- power off and login again -> success
- reset (by python script) removes the pin&credentials -> success
- lowering the number of flashpages to two throws an error after creating > 26 accounts -> success
- old accounts still work -> success

PeterKietzmann commented:
Think I would like three commits:
- flashpage stuff & recent fix
- reset command and script
- cleanup and documentation

What do you think?

PeterKietzmann commented on sys/include/fido2/ctap/ctap.h:
```diff
@@ -142,10 +141,9 @@ extern "C" {
 #else
 #define CONFIG_FIDO2_CTAP_DISABLE_UP 1
-#endif
+#endif /* BTN0_PIN */
```
What is actually the desired behavior? Silently disable user presence on boards that do not expose a pin? Or better throw a warning and exclude these boards from CI? Throw a warning here and disable user presence in the test by default?

Ollrogge replied:
I think we had a discussion about this when I added these lines and we agreed on the first approach. 
Murdock complained about a symbol being redefined. I hopefully fixed this issue now and it should work like the first approach with the recent changes.

PeterKietzmann replied:
Doesn't seem to be fixed. Maybe run the compile test locally before.

Ollrogge replied:
Sorry. Didn't think about just trying to compile locally with the failing boards. Did that now

Ollrogge replied:
Success :)

PeterKietzmann reviewed (approved):
All good. Green lights, go!
